src/sql-parser/src/ast/defs/query.rs

```diff
@@ -310,6 +310,65 @@ impl<T: AstInfo> AstDisplay for Distinct<T> {
     }
 }
 
+/// A block of common table expressions (CTEs).
+///
+/// The block can either be entirely "simple" (traditional SQL `WITH` block),
+/// or "mutually recursive", which introduce their bindings before the block
+/// and may result in mutually recursive definitions.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum CteBlock<T: AstInfo> {
+    Simple(Vec<Cte<T>>),
+    MutuallyRecursive(Vec<CteMutRec<T>>),
+}
+
+impl<T: AstInfo> CteBlock<T> {
+    /// Returns an empty (simple) CTE block.
+    pub fn empty() -> Self {
+        CteBlock::Simple(Vec::new())
+    }
+    /// True if there are no bindings in the block.
+    pub fn is_empty(&self) -> bool {
+        match self {
+            CteBlock::Simple(list) => list.is_empty(),
+            CteBlock::MutuallyRecursive(list) => list.is_empty(),
+        }
+    }
+    /// Returns a duplicate name if one exists. `None` means there are none.
+    pub fn reused_ident(&self) -> Option<&Ident> {
```

Comment:
Can do, but .. personal take is that `CteBlock` is a data structure and reporting duplicated identifiers isn't anything more than a method on that data structure. Choosing e.g. to panic if there are duplicates, or erroring in a builder, or to only support a map from `Ident` rather than a list would for sure be "more semantic".

But easy to copy/paste this in. I can't tell if "Can we ... ?" is "You must ..." or "Can we ... ?".

Reply:
Also, where would you put the free function? Part of the motivation is that it wants to be used in both `names.rs` and `query.rs`. If it were only either one, much more obvious to move it close to that use.

Reply:
Pondering a bit more, but I think what you might be saying is "could you move the call to `duplicated()` out of this method". Probably nothing wrong (I hope?) with enumerating the used identifiers; that's not too semantic is it? And then when the code does this:
```
    // A reused identifier indicates a reused name.
    if let Some(ident) = q.ctes.reused_ident() {
        sql_bail!(
            "WITH query name {} specified more than once",
            normalize::ident_ref(ident).quoted()
        )
    }
```
it could replace `reused_idents()` with `idents().duplicated()`. Or potentially put the `ident_ref` call there, rather than in the error.

Reply:
With with the last approach of the last comment.